[#5] 책 검색 라이브러리 연결

fried-shoes opened the pull request:
## 2025.08.06. Issue #5  에 대한 PR

1. Kakao API 연결
2. `ViewModel`, `Repository`, `data class` 생성
3. `ViewModel` - UI 연결
4. `ListFragment`의 `Toolbar` 에 검색어 표출
5. `BookListAdpater` 수정
6. 기타 코드 수정

fried-shoes committed:
feat: Kakao API 설정 준비 (의존성 추가)

- Retrofit: HTTP 통신 라이브러리
- Moshi: JSON 직렬화 라이브러리

fried-shoes committed:
feat: Kakao API & ViewModel 설정

- kakao key 를 로컬 파일에 숨기고 버전 관리 제외
- HTTP 요청 인터페이스 생성
- HTTP 요청에 대한 응답을 담을 data class 정의
- ViewModel, Repository 생성

fried-shoes committed:
feat: Toolbar 재설정

- 홈 화면에서는 툴바를 숨기고, 그 외에는 각 화면별로 title text 를 다르게 설정하기 위한 변경

fried-shoes committed:
feat: UI 에 ViewModel 붙이기

- Kakao API 를 이용한 책 목록 요청, 응답
- ListFragment toolbar title 설정

fried-shoes committed:
feat: BookListAdapter 및 item layout 수정

- 함수를 통해 book list 주입
- 책 설명 최대 두 줄로 제한, 줄임표 설정
- hint 속성 제거, tools 이용하여 개발할 때만 보이도록 설정

fried-shoes committed:
refactor: string resource 명명 변경

fried-shoes committed:
refactor: Activity, Fragment 패키지 이동

fried-shoes committed:
feat: RecyclerView 만으로도 스크롤 가능하므로 ScrollView 변경

fried-shoes committed:
feat: 현재 화면 id 에 따라 뒤로 버튼 클릭 설정

fried-shoes committed:
build: gradle, dependencies 버전 수정

fried-shoes committed:
chore: 코드 스타일링

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/adapter/BookListAdapter.kt:
```diff
@@ -3,18 +3,38 @@ package kr.yjkim.book_search.adapter
 import kr.yjkim.book_search.databinding.ItemBookListBinding
 
-class BookListAdapter(private val bookList: List<String>): RecyclerView.Adapter<BookListViewHolder>() {
+class BookListAdapter: RecyclerView.Adapter<BookListViewHolder>() {
```
`androidx.recyclerview.widget.ListAdapter<T, VH extends RecyclerView.ViewHolder>` 를 찾아서 한번 적용해보면 좋을 것 같습니다.

그럼 좀 더 효율적으로 동작할 수 있습니다!

fried-shoes replied:
`ListAdpater` 가 더 효율적인 이유는 아무래도 `DiffUtil` 클래스 때문일까요? 변경된 데이터만 업데이트하는...?

f-lab-tyler replied:
넵~! 맞습니다.

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/network/MoshiBuilder.kt:
```diff
@@ -0,0 +1,10 @@
+    val moshi: Moshi = Moshi.Builder()
+        .addLast(KotlinJsonAdapterFactory())
+        .build()
+}
```
따로 빼두는게 좋기는 하지만, 
Moshi 를 직접적으로 쓸 일이 거의 없기 때문에 `moshi` 객체는 `RetrofitClient` 로 옮기고 MoshiBuilder 는 제거하는게 어떨까 싶습니다 🙏

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/network/RetrofitClient.kt:
```diff
@@ -0,0 +1,14 @@
+        .baseUrl("https://dapi.kakao.com/")
+        .addConverterFactory(MoshiConverterFactory.create(moshi))
+        .build()
+        .create(KakaoService::class.java)
```
nit. Kotlin 에서는 inline 확장함수로 제공되는 create 가 있어서 요렇게 사용할 수 있습니다.

```kotlin
import retrofit2.create

// (중략)
.create<KakaoService>()
```

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/network/KakaoService.kt:
```diff
@@ -0,0 +1,17 @@
+
+interface KakaoService {
+
+    @Headers("Authorization: KakaoAK $REST_API_KEY")
```
요건 취향의 차이일 수 있는데, BuildConfig 가 static import 되어 있는데요.
저는 static import 는 좋아하지 않고, 아래처럼 클래스도 함께 명시하는걸 선호합니다.

이유는 코드 리뷰하는 동료들이 항상 pull 받아서 보는건 아니기 때문에,
Github 웹으로 볼 때도 명확할 필요가 있기 때문입니다.

클래스 명이 생략되어 있으면 import 를 고려하거나 이 값이 어디서 왔는지 찾아봐야하는데,
클래스 명이 명시되어 있으면 이를 인지하기 쉽습니다.
(코드 리뷰하는 동료 뿐만 아니라, 몇 일 뒤에 내가 코드를 다시 봐도 헷갈릴 때가 있기도 합니다 😅)

```kotlin
@Headers("Authorization: KakaoAK ${BuildConfig.REST_API_KEY}")
```

fried-shoes replied:
!!!! 안그래도 여러 `const val` 을 `import` 할 때 `import` 문이 너무 많고 길어져서 
'이걸 좀 묶을 방법이 없나' 생각한 적 있었는데, 이런 방법이 있었군요... 감사합니다! 
덤으로, 말씀해주신 것처럼 인지하기가 훨씬 쉽네요!

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/MyRepository.kt:
```diff
@@ -0,0 +1,29 @@
+import kr.yjkim.book_search.data.network.RetrofitClient
+import kr.yjkim.book_search.data.schema.BookItem
+
+class MyRepository {
```
이름에 검색에 대한 내용을 포함하고 검색만 하는 Repository 면 좋을 것 같습니다?!

fried-shoes replied:
음... 그럼 변수 `books` 의 `get/set` 을 없애고, `searchKeyword` 중단함수가 검색 결과를 리턴해서 `ViewModel` 에서 인지하도록 하면 괜찮을까요?

f-lab-tyler replied:
넵~ 그것도 좋은거 같습니다! 👍

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/MyRepository.kt:
```diff
@@ -0,0 +1,29 @@
+            } else {
+
+            }
+        } catch (e: Exception) {
```
예외 발생시 전부 잡으면 취소를 처리하기 어려워지기 때문에 CancellationException 은 다시 throw 해줘야합니다.

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/network/KakaoService.kt:
```diff
@@ -0,0 +1,17 @@
+    @GET("v3/search/book")
+    suspend fun getBookList(
+        @Query("query") keyword: String,
+    ): Response<BookResponse>
```
`Response` 가 꼭 필요한게 아니라면 BookResponse 를 바로 return 해줘도 됩니다.

fried-shoes replied:
`Response`를 빼면, 혹시 이 부분 (API 요청)이 네트워크 문제라던가 여러 가지 이유로 '실패'하게 되면 어떻게 캐치할 수 있나요...?

f-lab-tyler replied:
suspend 이기 때문에 네트워크 에러나 HTTP 에러 등이 발생하면 Exception 이 발생됩니다.

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/MyRepository.kt:
```diff
@@ -0,0 +1,29 @@
+            val response = kakaoService.getBookList(keyword)
+            if (response.isSuccessful) {
+                val result = response.body()!!
+                _books.value = result.bookList
```
`getBookList` 이 BookResponse 를 바로 반환하면 문제가 해소될거 같긴 한데,
이런 경우에는 `!!` 보다 `requireNotNull` 함수를 사용해서 명확히 이 로직상으로는 null 이 될 수 없다는걸 표현하는게 좋습니다.

```kotlin
val result = requireNotNull(response.body())
```

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/HomeFragment.kt:
```diff
@@ -1,19 +1,22 @@
 
     private var _binding: FragmentHomeBinding? = null
     private val binding get() = _binding!!
+    private val viewModel: MyViewModel by activityViewModels()
```
요건 `activityViewModels` 롤 꼭 둬야하는거...죠?!

이왕이면 역할에 따라 ViewModel 을 분리하고, 
Fragment 끼리 공유할 필요가 없다면 Fragment 의 생명 주기를 가지는 ViewModel 을 다루는게 좋을 것 같아서 여쭤봅니다!

fried-shoes replied:
음... `HomeFragment` 에서 검색한 검색어를 `ListFragment`에 사용하고, `ListFragment` 에서 선택한 책의 제목을 `InfoFragment` 의 `toolbar title`로 사용할 예정이라... 
제 짧은 생각으로는 `ViewModel`을 공유하는 것 말고 다른 방법이 떠오르지 않아서요...ㅠ

f-lab-tyler replied:
가능하다면 분리하는게 좋겠다 인데 힘들다면 `activityViewModels ` 이어도 괜찮습니다.

다만, 데이터를 주고받는 것과 공유하는건 다르기 때문에,
단순히 데이터를 주고 받는 것으로 해결된다면 argument 로 넘기거나 [FragmentResultListener](https://developer.android.com/guide/fragments/communicate?hl=ko#fragment-result) 로 결과를 받을 수도 있어서,
이 부분도 한번 고민은 해보면 좋겠다는 의도였습니다 🙏

fried-shoes replied:
아하! 넵 한번 도전해보겠습니다! 💪

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/ListFragment.kt:
```diff
@@ -15,22 +16,26 @@ class ListFragment: Fragment() {
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
         _binding = FragmentListBinding.inflate(inflater, container, false)
+
+        val toolbarTitleText = getString(R.string.toolbar_list_title, vm.keyword.toString())
```
nit. `vm.keyword.toString()` 에서 toString 은 필요없다고 경고가 뜨는군요 😄

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/MyViewModel.kt:
```diff
@@ -0,0 +1,33 @@
+        }
+        throw IllegalArgumentException("Unknown ViewModel class")
+    }
+}
```
nit. MyViewModel 에 이렇게 정의하는 것도 방법입니다.

```kotlin
companion object {
    fun create(repository: MyRepository): ViewModelProvider.Factory {
        return viewModelFactory {
            initializer {
                MyViewModel(repository)
            }
        }
    }
}
```

fried-shoes committed:
refactor: moshi 객체 이동, MoshiBuilder 제거

fried-shoes committed:
refactor: create() 를 kotlin 확장 inline 함수로 변경

fried-shoes committed:
refactor: BuildConfig 클래스 명시

fried-shoes committed:
refactor: toString() 불필요, 삭제

fried-shoes committed:
refactor: ViewModelProvider.Factory 다르게 정의하기

fried-shoes committed:
feat: BookListAdapter 를 ListAdapter<T, VH> 확장하도록 변경

- RecyclerView.Adapter<VH> 보다 ListAdapter<T, VH> 가 더 효율적
- 매번 전체 리스트를 갱신하는 것이 아니라, 변경된 아이템만 갱신함

fried-shoes committed:
refactor: Repository 이름 변경

fried-shoes committed:
feat: Repository 가 '검색' 관련 기능만 수행하도록 수정

- List<BookItem> get/set 은 ViewModel 로 이동

fried-shoes committed:
feat: CancellationException 분기 추가

- 코루틴이 취소될 경우를 처리하기 위해 반드시 CancellationException 을 catch 및 throw 할 것

fried-shoes committed:
feat: Retrofit Interface 함수에서 Response 제거

fried-shoes committed:
feat: Safe Args 사용하여 Fragment 간 검색 키워드 전달

fried-shoes committed:
feat: Fragment 별 ViewModel 분리

- BookSearchRepository 를 공유하기 위해 Object로 변경
- 각각 ViewModel 을 가지게 되면 사실상 MainActivity 는 ViewModel 을 사용하지 않으므로 레이아웃도 수정

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/BookSearchRepository.kt:
```diff
@@ -0,0 +1,27 @@
+        } catch (e: CancellationException) {
+            throw e
+        } catch (_: Exception) {
+            emptyList()
```
여기서 빈 리스트를 반환하면 사용하는 곳에서는 `searchKeyword` 의 실패를 어떻게 전달받을 수 있을까요?

fried-shoes replied:
음... 그럼 일반 예외(CancellationException 을 제외한 예외)도 역시 throw 해서 ViewModel 에서 처리하도록 하는 게 더 나은 구조일까요? 
일단 저는... 실패하면 일단 리스트가 안 나오는 게 맞으니까 null 은 안되고 emptyList() 를 넣고,
뭔가 문제가 생긴 걸 사용자에게 알리기는 해야 하니까, 한다면 다이얼로그 또는 레이아웃에서 '결과 없음' 같은 텍스트를 보여주는 쪽으로 생각했습니다. 

혹시 emptyList() 반환하는 부분이 catch 블록임에도 불구하고 viewModelScope.launch 에서 예외로 인식하지 않는 것일까요? 😢

f-lab-tyler replied:
`emptyList()` 를 반환하면 ViewModel 은 실제로 값이 비어있는건지, 에러로 인해 값이 비어있는건지 모호해집니다.
그래서 몇가지 방법들이 있겠지만 중요한건 _어디에서 예외를 잡아서 처리할 것이냐_ 를 정의하는 것입니다.

우리는 UI 를 다루고 있기 때문에, 결국 무엇인가 UI 로 표현할 가능성이 높습니다.
그래서 전 Repository 는 예외를 있는 그대로 던지고 ViewModel 에서 이를 잡아서 처리하는걸 선호합니다.

예를 들어 Repository 에서 예외를 던지고 ViewModel 에서 이를 잡아서 `재시도 버튼이 포함된 에러 페이지` 같은걸 보여줄 수도 있겠죠.

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/ListViewModel.kt:
```diff
@@ -0,0 +1,22 @@
+
+class ListViewModel(repository: BookSearchRepository): ViewModel() {
+
+    val bookList = repository.bookList
```
이런 경우에는 타입을 명시하는 편이 좋습니다.
왜냐하면 타입을 보고 싶으면 안드로이드 스튜디오에서 확인하거나, BookSearchRepository 로 들어가야 볼 수 있기 때문입니다.

코드는 최대한 이 파일 안에서 최대한 명확하게 표현할 수 있어야 합니다.
생략할 수 있다고 해서 생략하면 코드의 가독성이 떨어질 수도 있습니다.

```kotlin
val bookList: LiveData<List<BookItem>> = repository.bookList
```

f-lab-tyler reviewed (approved):
소소한 코멘트를 남겨봤습니다. 나머지는 괜찮은듯 해서 미리 Approve 드립니다.

fried-shoes committed:
refactor: 객체 타입 명시
